### ISSUE TITLE
[sBTC DR] Architecture and roadmap

### ISSUE THREAD

**Issue by stjepangolemac** [stacks-network/sbtc#57]
After a long session with @netrome and @jferrant we've come up with an architecture and roadmap that we think (and hope) should produce a reliable system within the time restrictions we have.

We propose the new system to be event sourced, and its logic to be split into actors. This approach seems to be fit for the problem of keeping track of "financial like" data over multiple sources of truth.

The fundamental concept is the Event type, which clearly shows the complexity of the process.

```rust
enum Event {
  AssetContractDeployed
  DepositSeen
  MintCreated
  MintBroadcasted
  MintConfirmed
  MintRejected
  WithdrawalSeen
  BurnCreated
  BurnBroadcasted
  BurnConfirmed
  BurnRejected
  FulfillmentCreated
  FulfillmentBroadcasted
  FulfillmentConfirmed
  FulfillmentRejected
}
```

From these variants we've been able to assign different roles for the components of the system. We call every component an actor and structure them as explained below.

```text
- IO Actors
    - BitcoinObserver
        - writes
            - DepositSeen
            - WithdrawalSeen
            - FulfillmentSeen
    - BitcoinWriter
        - reads
            - FulfillmentCreated
        - writes
            - FulfillmentBroadcasted
    - StacksWriter
        - reads
            - MintCreated
            - BurnCreated
        - writes
            - MintBroadcasted
            - BurnBroadcasted
    - StacksObserver
        - writes
            - MintConfirmed
            - MintRejected
            - BurnConfirmed
            - BurnRejected
- Processing actors
    - DepositProcessor
        - reads
            - DepositSeen
            - MintBroadcasted
            - MintConfirmed
            - MintRejected
        - writes
            - MintCreated
    - WithdrawalBurnProcessor
        - reads
            - WithdrawalSeen
            - BurnBroadcasted
            - BurnConfirmed
            - BurnRejected
        - writes
            - BurnCreated
    - WithdrawalFulfillmentProcessor
        - reads
            - DepositSeen
            - BurnConfirmed
            - FulfillmentBroadcasted
            - FulfillmentConfirmed
            - FulfillmetnRejected
        - writes
            - FulfillmentCreated
```

We think this work can be broken up into these tasks:

- [x] #58
- [x] #59
- [x] #60
- [x] #61
- [x] #62
- [x] #63
- [x] #64
- [x] #65
- [x] #66
- [x] #67
- [x] #68
- [x] #69
- [x] #70
- [x] #71
- [x] #73

**Comment by jferrant** [comment]
Is so nice that we don't need to have a handoff/distributed signing in DR. :)

**Comment by friedger** [comment]
All features implemented for 0.1 and on `main`
